### ISSUE TITLE
Threads hanging consuming a full core

### ISSUE THREAD

**Issue by Nemo157** [rust-lang/docs.rs#1942]
First noticed 2022-11-14 we had much higher cpu usage than usual

![image](https://user-images.githubusercontent.com/81079/205119582-42619390-32bc-462e-b39d-29e63a3411d8.png)

The `cratesfyi` process was at 300% cpu usage. After restarting the process everything went back to normal.

Again on 2022-11-30 we noticed high cpu usage, this time checking the per-thread stats it was 2 unnamed threads causing it. Again a restart fixed it

I suspected it was related to #1842, so ran a script highlighting every readme from the production database dump, that managed to find one readme that hung at 100% cpu usage. Opened an upstream issue https://github.com/trishume/syntect/issues/460, but this is likely not the only path through which this can be triggered.

**Comment by Nemo157** [comment]
Looks like there might still be something triggering this, we just got slow requests again and had 4 render threads at high cpu usage

![image](https://github.com/rust-lang/docs.rs/assets/81079/6844d5e2-aa82-420e-9626-86e0539e2e1c)

**Comment by Nemo157** [comment]
Yep, looking through early request failures during this last outage I found one which hangs my local server. Will try and take a look at whether there's something better we can do to avoid this issue.

**Comment by syphar** [comment]
Thinking about this one I would close this issue. 

There were some safeguards implemented, so the risk of this happening again should be lower. 

Of course due to the regex engine, user generated content, and the syntax highlighting engine, we still have a risk of this happening, probably only possible to prevent by using a pool of subprocesses for doing this rendering, where we can kill based on timeouts. 
Or going back to frontend rendering.
